das/algorithms/3DES.py

```diff
@@ -0,0 +1,324 @@
+        for j in range(1, len(bit_string)):
+            s = s + bit_string[j]
+        s = s + bit_string[0]
+        bit_string = s
+        s = ""
+    return bit_string
+
+
+def RS(bit_string, num_of_shifts):
+    for i in range(num_of_shifts):
+        bit_string = bit_string[-1] + bit_string[:-1]
+
+    return bit_string
+
+
+def xor(a, b):
+    result = ""
+    for i in range(len(a)):
+        if a[i] == b[i]:
+            result = result + "0"
+        else:
+            result = result + "1"
+    return result
+
+
+def get_round_key_encrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(16):
+        left_key = LS(left_key, shift_table[round_count])
+        right_key = LS(right_key, shift_table[round_count])
+
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
+
+    return round_keys
+
+
+def get_round_keys_decrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(15,-1,-1):
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
+
+        left_key = RS(left_key, shift_table[round_count])
+        right_key = RS(right_key, shift_table[round_count])
+    return round_keys
+
+
+def Expansion(binary):
+    return "".join([binary[i-1] for i in exp_d])
+
+
+def Permutation(binary):
+    return "".join([binary[i-1] for i in per])
+
+
+def S(binary):
+    inputs = [binary[i:i+6] for i in range(0,48,6)]
+    coords = [(bin2dec(x[0]+x[-1]), bin2dec(x[1:-1])) for x in inputs]
+    outputs = [dec2bin(sbox[i][x[0]][x[1]]) for i, x in enumerate(coords)]
+    return "".join(outputs)
+
+
+def F(right, k):
+    tmp = Expansion(right)
+    tmp = xor(tmp, k)
+    tmp = S(tmp)
+    tmp = Permutation(tmp)
+    return tmp
+
+
+def encrypt(pt, key):
+    assert len(pt) == 64
+
+    left_pt = pt[0:32]
+    right_pt = pt[32:]
+
+    for k in get_round_key_encrypt(key):
+        left_pt, right_pt = right_pt, xor(left_pt, F(right_pt, k))
+
+    return right_pt + left_pt
+
+
+def decrypt(ct, key):
+    left_pt = ct[0:32]
+    right_pt = ct[32:]
+
+    for k in get_round_keys_decrypt(key):
+        left_pt, right_pt = right_pt, xor(left_pt, F(right_pt, k))
+
+    return right_pt + left_pt
+
+
+def triple_des_encrypt(pt, key1, key2, key3):
+    if not key1 == key2 and not key3 == key2:
+        raise Exception("key2 must equal key1 or key3")
+    ct = encrypt(pt, key1)
+    ct = decrypt(ct, key2)
+    ct = encrypt(ct, key3)
+    return ct
+
+
+def triple_des_decrypt(ct, key1, key2, key3):
+    if not key1 == key2 and not key3 == key2:
+        raise Exception("key2 must equal key1 or key3")
+    pt = decrypt(ct, key1)
+    pt = encrypt(pt, key2)
+    pt = decrypt(pt, key3)
```

Comment:
@Lionion Should the keys be traversed in reverse for decrypt? This removes the requirement that key2 is a duplicate